twilight-gateway/src/shard.rs

```diff
@@ -0,0 +1,987 @@
+    /// [`CloseCode`] for possible reasons.
+    ///
+    /// [failed authentication]: CloseCode::AuthenticationFailed
+    /// [invalid intents]: CloseCode::InvalidIntents
+    FatallyClosed {
+        /// Close code of the close message.
+        ///
+        /// The close code may be able to parse into [`CloseCode`] if it's a
+        /// known close code. Unknown close codes aren't considered fatal.
+        close_code: u16,
+    },
+}
+
+impl ConnectionStatus {
+    /// Determine the connection status from the close code.
+    ///
+    /// Defers to [`CloseCode::can_reconnect`] to determine whether the
+    /// connection can be reconnected, defaulting to [`Self::FatallyClosed`] if
+    /// the close code is unknown.
+    fn from_close_frame(maybe_frame: Option<&CloseFrame<'static>>) -> Self {
+        let raw_code = if let Some(frame) = maybe_frame {
+            frame.code()
+        } else {
+            return Self::Disconnected {
+                close_code: None,
+                reconnect_attempts: 0,
+            };
+        };
+
+        let can_reconnect = CloseCode::try_from(raw_code)
+            .map(CloseCode::can_reconnect)
+            .unwrap_or(true);
+
+        if can_reconnect {
+            Self::Disconnected {
+                close_code: Some(raw_code),
+                reconnect_attempts: 0,
+            }
+        } else {
+            Self::FatallyClosed {
+                close_code: raw_code,
+            }
+        }
+    }
+
+    /// Whether the shard is connected.
+    pub const fn is_connected(self) -> bool {
+        matches!(self, Self::Connected)
+    }
+
+    /// Whether the shard has disconnected but may reconnect in the future.
+    pub const fn is_disconnected(self) -> bool {
+        matches!(self, Self::Disconnected { .. })
+    }
+
+    /// Whether the shard has fatally closed, such as due to an invalid token.
+    pub const fn is_fatally_closed(self) -> bool {
+        matches!(self, Self::FatallyClosed { .. })
+    }
+}
+
+/// Gateway event with only minimal required data.
+#[derive(Deserialize)]
+struct MinimalEvent<T> {
+    /// Attached data of the gateway event.
+    #[serde(rename = "d")]
+    data: T,
+}
+
+/// Minimal [`Ready`] for light deserialization.
+///
+/// [`Ready`]: twilight_model::gateway::payload::incoming::Ready
+#[derive(Deserialize)]
+struct MinimalReady {
+    /// ID of the new identified session.
+    session_id: String,
+}
+
+/// Shard to run and manage a session with the gateway.
+///
+/// Shards are responsible for handling incoming events, process events relevant
+/// to the operation of shards - such as requests from the gateway to re-connect
+/// or invalidate a session - and then pass the events on to the user via an
+/// event stream.
+///
+/// Shards will [go through a queue][`queue`] to initialize new ratelimited
+/// sessions with the ratelimit. Refer to Discord's [documentation][docs:shards]
```

Comment:
```suggestion
/// session with the ratelimit. Refer to Discord's [documentation][docs:shards]
```

Reply:
The plural form of session is correct here